### PR TITLE
rgw: add an option to recalculate user stats

### PR DESCRIPTION
**Pull request by theanalyst**
This series of commits adds an option to radosgw-admin called recalculate-stats
which resets the user head object to stats summed up according to user.bucket
omap entries. We call sync stats after reseting the stats in rgw-admin so that
stats are in sync with the current time.

cls/user: Introducing reset user stats which rewrites the user.buckets header to
the sum of entries and sizes from bucket omap entries

### CODE REVIEW

**Comment by theanalyst** [comment]
Another possibility was adding yet another flag at `cls_user_set_buckets_info` for reset and checking that before we decrement stats,  but that is making the function more complicated than it already is.

File: src/rgw/rgw_rados.h [hunk 1]
```diff
@@ -3613,6 +3613,7 @@ class RGWRados : public AdminSocketHook
   int fix_tail_obj_locator(const RGWBucketInfo& bucket_info, rgw_obj_key& key, bool fix, bool *need_fix);
 
   int cls_user_get_header(const string& user_id, cls_user_header *header);
+  int cls_user_clear_header(const string& user_id);
```
Comment (oritwas) [inline src/rgw/rgw_rados.h]: I prefer cls_user_recalc_stats or cls_user_reset_stats

File: src/rgw/rgw_rados.cc [hunk 2]
```diff
@@ -13272,6 +13272,28 @@ int RGWRados::cls_user_get_header(const string& user_id, cls_user_header *header
...
+  int r = get_raw_obj_ref(obj, &ref);
+  if (r < 0) {
+    return r;
+  }
+
+  librados::ObjectWriteOperation op;
+  ::cls_user_reset_stats(op);
+  r = ref.ioctx.operate(ref.oid, &op);
+
+  if (r < 0)
```
Comment (oritwas) [inline src/rgw/rgw_rados.cc]: return r will be cleaner

File: src/rgw/rgw_rados.h (hunk 1 above)
Comment (theanalyst) [reply src/rgw/rgw_rados.h]: sorry that was an initial method forgot while refactoring, fixing.

File: src/rgw/rgw_admin.cc [hunk 3]
```diff
@@ -6140,6 +6143,25 @@ int main(int argc, const char **argv)
...
+      if (!bucket_name.empty()){
+	cerr << "ERROR: recalculate doesn't work on buckets" << std::endl;
+	return EINVAL;
+      }
+      ret = store->cls_user_clear_header(user_str);
+      if (ret < 0) {
+	cerr << "ERROR: could not clear user stats: " << cpp_strerror(-ret) << std::endl;
+	return -ret;
+      }
+      sync_stats = true;
```
Comment (theanalyst) [inline src/rgw/rgw_admin.cc]: @oritwas I can drop the sync_stats here, so the user can call it with --reset-stats --sync-stats if they need to explicitly sync stats or should we make sure that reset stats are also in sync with what cluster sees

Comment (oritwas) [reply src/rgw/rgw_admin.cc]: you can drop it as you don't need it because you write the heade with the recalculated stats.

**Comment by theanalyst** [comment]
Changelog: 
- rename admin cli option to reset stats 
- rename rgw rados methods to user_reset_stats
- add tracker ref http://tracker.ceph.com/issues/23335

**Comment by theanalyst** [comment]
@oritwas do the updates look ok?

**Comment by smithfarm** [comment]
@theanalyst @yuriw This PR is still open, yet its luminous backport [1] was merged?

[1] https://github.com/ceph/ceph/pull/21393

**Comment by theanalyst** [comment]
we discussed this at standups yesterday and was given signal to merge

**Pull request merged**